util/syncutil/once.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2016 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Nathan VanBenschoten (nvanbenschoten@gmail.com)
+
+package syncutil
+
+import "sync/atomic"
+
+// Once is an object that will perform exactly one action.
+//
+// This object behaves the same as sync.Once from the standard library, except
+// that it holds a reference to a Mutex instead of holding the Mutex by value.
+// This means that it can be used in conjunction with external synchronization
+// that uses the same Mutex.
+type Once struct {
```

Comment:
The types in `syncutil` that have the same names as ones from `sync` should be drop-in replacements. This isn't, because it requires the mutex to be initialized explicitly. 

Why does it matter that the lock in `Once` be the same as an external lock? Wouldn't it be the same to use a normal `sync.Once` and then acquire the external lock inside the callback?